libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -4309,6 +4312,23 @@ auto WaveTrack::SortedIntervalArray() const -> IntervalConstHolders
    return result;
 }
 
+bool WaveTrack::ClipsAreUnique() const
+{
+   constexpr auto comp =
+   [](const WaveClipHolder &x, const WaveClipHolder &y){
+      return x.owner_before(y);
+   };
+   constexpr auto equivalent =
+   [comp](const WaveClipHolder &x, const WaveClipHolder &y){
+      return !(comp(x, y) || comp(y, x));
+   };
+   auto pointers = mClips;
+   const auto b = begin(pointers),
+      e = end(pointers);
+   sort(b, e, comp);
+   return adjacent_find(b, e, equivalent) == e;
```

Comment:
I'm having trouble understanding this:
`!(comp(x, y) || comp(y, x))` returns false if x and y are shared pointers to the same object, so if there are as many different clip addresses as there are shared pointers, all comparisons in the range `[b, e)` will return `false`. Won't then `adjacent_find` return an iterator to `b` ?

What about the following?
```
   std::unordered_set<const WaveClip*> addresses;
   for (const auto &clip : mClips)
      addresses.insert(clip.get());
   return addresses.size() == mClips.size();
   ```

Reply:
> I'm having trouble understanding this: `!(comp(x, y) || comp(y, x))` returns false if x and y are shared pointers to the same object, so if there are as many different clip addresses as there are shared pointers, all comparisons in the range `[b, e)` will return `false`. Won't then `adjacent_find` return an iterator to `b` ?
> 
> What about the following?
> 
> ```
>    std::unordered_set<const WaveClip*> addresses;
>    for (const auto &clip : mClips)
>       addresses.insert(clip.get());
>    return addresses.size() == mClips.size();
> ```

No,  I wrote the equivalence correctly, as not-inequivalent.  If x is less than y or y is less than x, they are inequivalent.

But indeed I could write all this more simply, if not more efficiently.  But the efficiency is not important.

For some reason I thought operator < was not defined for shared_ptr but I'm wrong about that, so a simple std::set should work, but it will do a lot of heap allocation of tree nodes.

See also https://en.cppreference.com/w/cpp/header/flat_set